pkg/kv/kvserver/kvflowcontrol/kvflowconnectedstream/stream.go

```diff
@@ -0,0 +1,1339 @@
+// of Next is protocol state internal to raft, regarding what to retry
+// sending. NextUpperBound can regress if the replica transitions out of
+// StateReplicate, and back into StateReplicate.
+//
+// The Ready described here is only the subset that is needed by replication
+// AC -- heartbeats, appending to the local log etc. are not relevant.
+//
+// Ready must be called on every tick/ready of Raft since we cannot tolerate
+// state transitions from
+//
+// StateReplicate => {state in all states: state != StateReplicate} =>
+// StateReplicate
+//
+// that is not observed by RangeController. Since outgoing messages are sent
+// in the tick/ready handling and a return from the intermediate state to
+// StateReplicate relies on receiving responses to those outgoing messages, we
+// should not miss any transitions. If stale messages received in a step can
+// cause transitions out and back without observation, we can add a monotonic
+// counter for each follower inside Raft (this is just local state at the
+// leader), which will be incremented on every state transition and expose
+// that via the RaftInterface.
+type Ready interface {
+	// Entries represents the new entries that are being added to the log.
+	// This may be empty.
+	//
+	// The second byte of AC entries will encode the priority, so we will have
+	// the priority, size, and index position, which is all that we need. Unlike
+	// v1, where we are doing the accounting in
+	// https://github.com/cockroachdb/cockroach/blob/f601b7b439ced71030bfdb0d9ba9cb4925420569/pkg/kv/kvserver/replica_proposal_buf.go#L1057-L1066
+	// which is messy.
+	Entries() []raftpb.Entry
+	// RetransmitMsgApps returns the MsgApps that are being retransmitted, or
+	// being used to ping the follower. These will never be queued by
+	// replication AC, but it may adjust the priority before sending (due to
+	// priority inheritance, when the Message has non-empty Entries).
+	RetransmitMsgApps() []raftpb.Message
+}
+
+// RaftInterface abstracts what the RangeController needs from the raft
+// package, when running at the leader. It also provides one piece of
+// information that is not in the raft package -- the current connected state
+// of the store, that we will use the RaftTransport to answer.
+//
+// NB: group membership and connectivity information is communicated to the
+// RangeController via a separate channel, as in the v1 implementation, since
+// it depends on information inside and outside Raft. The methods in Raft are
+// the ones RangeController will call.
+//
+// The implementation should not need to acquire Replica.mu since we have not
+// said anything about the relative lock ordering of Replica.mu and the
+// internal mutexes in RangeController.
+type RaftInterface interface {
+	// FollowerState returns the current state of a follower. The value of
+	// nextUpperBound is populated iff in StateReplicate. All entries >=
+	// nextUpperBound have not yet had MsgApps constructed.
```

Comment:
In case it impacts correctness reasoning somewhere, "have not yet had MsgApps constructed" has an implicit asterisk: "by this connected stream / StateReplicate" (at some concrete in-memory "epoch" if we number them).

There can be in-flight `MsgApp`s at higher indices, and we may even get acks for them (beyond the `Next` index). E.g. if we just transitioned out / back in `StateReplicate` and lost some tracking information.